Global cleanup, MC 1.17, Java 11 and repackaged FAWE.

Aurelien30000 opened the pull request:
Apart from a standard version update, this PR completely  removes use of former and unsafe _Async_ classes, and take advantage of edit sessions and WE alternatives!
Overall, the plugin has been reworked especially for FAWE and improved on some points.

Aurelien30000 committed:
Small cleanup, MC 1.17, Java 16 and repackaged FAWE.

Aurelien30000 committed:
Small improvement

NotMyFault commented on .github/workflows/build.yml:
```diff
@@ -6,7 +6,7 @@ jobs:
-        java: ["8", "11"]
+        java: ["16"]
         os: ["ubuntu-latest"]
     runs-on: "${{ matrix.os }}"
```
That can be minified to a couple of lines `16` and `ubuntu-latest`. No need to use a matrix for a single value.

Aurelien30000 replied:
Right, ty.

NotMyFault commented on build.gradle.kts:
```diff
@@ -9,7 +8,7 @@ plugins {
-	sourceCompatibility = JavaVersion.VERSION_1_8
+	sourceCompatibility = JavaVersion.VERSION_16
 	targetCompatibility = sourceCompatibility
 }
```
That is deprecated as of Gradle 7.1. We can use `JavaPluginExtension` here. Like 
```kt
the<JavaPluginExtension>().toolchain {
    languageVersion.set(JavaLanguageVersion.of(16))
}
```
for example, but set the output to 11.

Aurelien30000 replied:
Sure, ty.

NotMyFault commented on build.gradle.kts:
```diff
@@ -21,16 +20,16 @@ repositories {
 
 group = "com.thevoxelbox"
-version = "1.0.4-backward"
+version = "1.0.5-backward"
```
This is a minor version change so that'd be 1.1.0-backward

Aurelien30000 replied:
Wasn't sure, ty.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/Favs.java:
```diff
@@ -16,60 +16,62 @@
 
 public class Favs {
-    private final VoxelSniperPlugin plugin;
+	private final VoxelSniperPlugin plugin;
```
Putting that here for the all classes, reformatting/whitespaces is not needed.

Aurelien30000 replied:
As you want ;)

Aurelien30000 replied:
**Note:** ``.editorconfig`` file is configured to use tabs as indentation character.

Aurelien30000 replied:
Most of the classes use this identation _(tabs)_, imo this change should persist for consistency.

NotMyFault replied:
That is a truly horrific indent style. Keep it for now, I will apply our org editorconfig later on.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/wrapper/AsyncWorld.java:
```diff
@@ -1250,6 +1226,71 @@ public Location locateNearestStructure(@NotNull Location arg0, @NotNull Structur
+	public @NotNull Collection<Material> getInfiniburn() {
+		return parent.getInfiniburn();
+	}
+
```
Unneeded and undesired. The Async* classes have been removed from Fawe since they are inherently unsafe and not recommendable to be used. I originally copied them here, because I didn't have the motivation to add it properly.
The proper way would be what Fawe does; use an EditSession instead of an AsyncWorld same should be done here and yeetus deletus all async classes.

Aurelien30000 replied:
I'll take a look, some help with understanding EditSessions would be appreciated _(how to use them to follow your code styles and FAWE?)_.

Instead, what should I do, return null for every method, or an unsupported exception?

NotMyFault replied:
> Instead, what should I do, return null for every method, or an unsupported exception?

Yeetus deletus alongside the async classes. The override is needed, because these classes used to replace native methods delegating to our own implementation. Using an ES and deleting the classes will get a rid of the overrides

> I'll take a look, some help with understanding EditSessions would be appreciated (how to use them to follow your code styles and FAWE?).

Put them in a try with resources block, so they are closed properly when the the code exits the block, looks always a bit cleaner. But for a real "style" take a look at the Fawe code for ES's; our checkstyle plugin is not in use atm so no biggie but make sure it doesn't look out of place.

Aurelien30000 replied:
Pheew, this creepy _migration_ is hopefully done, everything is now based on an editsession. I am committing the changes tomorrow, after a proofreading and tests.

Aurelien30000 replied:
This request has been resolved and several amateur players have tried it 😅 Hopefully nothing broke.

**Note about undo code**: Since FAWE completely handles undo on its own side, may I _remove_/_comment out_ all undo related code?
Also, with such changes, wouldn't a major version like ``2.0.0``, you suggested in another comment, be more adequate?

NotMyFault replied:
> Pheew, this creepy migration is hopefully done, everything is now based on an editsession. I am committing the changes tomorrow, after a proofreading and tests.

I'll give it a read at a later point.

> Note about undo code: Since FAWE completely handles undo on its own side, may I remove/comment out all undo related code?

Since we're now using an ES, undo/redo/history at all can be handled thru Fawe. Letting Fawe handle that makes named commands obsolete anyway.

> Also, with such changes, wouldn't a major version like 2.0.0, you suggested in another comment, be more adequate?

Likely. I created an additional branch, `feature/1.17.1` you can target this pr on allowing us to address several things and merging it all in one if ready.

NotMyFault commented on build.gradle.kts:
```diff
@@ -21,16 +20,16 @@ repositories {
+	compileOnlyApi("io.papermc.paper:paper-api:1.17-R0.1-SNAPSHOT")
 	compileOnly("org.jetbrains:annotations:20.1.0")
-	compileOnlyApi("com.intellectualsites.fawe:FAWE-Bukkit:1.16-637")
+	compileOnlyApi("com.fastasyncworldedit:FAWE-Bukkit:1.17-23")
```
1.17-44 & use https://mvn.intellectualsites.com/content/groups/public/ as maven repository replacing the current two repositories from us.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/CopyPastaBrush.java:
```diff
@@ -136,21 +136,13 @@ private void doPasta(Snipe snipe) {
+						case 90 -> clampY(this.pastePoint[0] - this.offsetPoint[2] - k, this.pastePoint[1] + this.offsetPoint[1] + j, this.pastePoint[2] + this.offsetPoint[0] + i);
+						default -> // assume no rotation
+							clampY(this.pastePoint[0] + this.offsetPoint[0] + i, this.pastePoint[1] + this.offsetPoint[1] + j, this.pastePoint[2] + this.offsetPoint[2] + k);
+					};
```
Putting this here one time for all classes; using Java 16 features are undesired at this time. Since Fawe still uses a minimum Java version of 11 and I see no good reason to require Java 16 here but for sake of using enhanced switches, 11 will remain.

Aurelien30000 committed:
Resolve some requests

NotMyFault commented on .github/workflows/build.yml:
```diff
@@ -4,18 +4,14 @@ on: ["pull_request", "push"]
         with:
           distribution: "adopt"
-          java-version: "${{ matrix.java }}"
+          java-version: "11"
```
This step can use 16 like outlined in my previous comment.
It sets the Java version of the GitHub action runner, but has no impact on the Java version of the project.

NotMyFault commented on build.gradle.kts:
```diff
@@ -8,28 +8,34 @@ plugins {
+	options.release.set(11)
+}
+
+configurations.all {
+	attributes.attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, 16)
+}
```
Unneeded like outlined in my previous comment. We don't need to enforce Java 16 for no good reason.

Aurelien30000 replied:
Humm, without this setting, build fails:

```java
Execution failed for task ':compileJava'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve io.papermc.paper:paper-api:1.17.1-R0.1-SNAPSHOT.
     Required by:
         project :
      > No matching variant of io.papermc.paper:paper-api:1.17.1-R0.1-SNAPSHOT:20210707.082846-3 was found. The consumer was configured to find an API of a library compatible with Java 11, preferably in the form of class files, preferably optimized for standard JVMs, and its dependencies declared externally but:
          - Variant 'apiElements' capability io.papermc.paper:paper-api:1.17.1-R0.1-SNAPSHOT declares an API of a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component compatible with Java 16 and the consumer needed a component compatible with Java 11
              - Other compatible attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
          - Variant 'javadocElements' capability io.papermc.paper:paper-api:1.17.1-R0.1-SNAPSHOT declares a runtime of a component, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 11)
                  - Doesn't say anything about its elements (required them preferably in the form of class files)
          - Variant 'runtimeElements' capability io.papermc.paper:paper-api:1.17.1-R0.1-SNAPSHOT declares a runtime of a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component compatible with Java 16 and the consumer needed a component compatible with Java 11
              - Other compatible attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
          - Variant 'sourcesElements' capability io.papermc.paper:paper-api:1.17.1-R0.1-SNAPSHOT declares a runtime of a component, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 11)
                  - Doesn't say anything about its elements (required them preferably in the form of class files)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.


```

NotMyFault replied:
That'll need to stay then.

Aurelien30000 committed:
Re-add JDK 16 for actions

Aurelien30000 committed:
Bump serverlib

Aurelien30000 committed:
Global cleanup, delete async classes and most accesses to bukkit api, through our implementation.
Use editsession and WE alternatives instead.

Aurelien30000 committed:
Bump major version => 2.0.0

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/PullBrush.java:
```diff
@@ -247,5 +233,13 @@ public int getY() {
+			return this.blockData;
+		}
+
+		public double getStr() {
+			return this.str;
```
Please use a sane name for this method. This is somewhat of unintuitive if you don't know what you are looking for.

Aurelien30000 replied:
Certainly, but I have absolutely no idea about the returned result provided by the following method...

```java
private double getStr(double t) {
		double lt = 1 - t;
		return (lt * lt * lt) + 3 * (lt * lt) * t * this.c1 + 3 * lt * (t * t) * this.c2; // My + (t * ((By + (t * ((c2 + (t * (0 - c2))) - By))) - My));
}
```

NotMyFault replied:
It's crystal clear what it does and therefore the method naming will remain.

Matocolotoe replied:
No it's not

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SignOverwriteBrush.java:
```diff
@@ -335,6 +348,40 @@ private void resetStates() {
+			case BlockID.JUNGLE_SIGN:
+			case BlockID.JUNGLE_WALL_SIGN:
+			case BlockID.OAK_SIGN:
+			case BlockID.OAK_WALL_SIGN:
```
The sign names need to be escaped. *_wall stances didn't exist in 1.13 which we still support. Escaping them also allows adding all "new" sign materials and not failing to load on older versions.

Aurelien30000 replied:
Wdym exactly, using their names instead of a defined and incomplete enum in 1.13, as your pointed out?

NotMyFault replied:
Ignore my previous comment. After taking a look at the code where named method is used, it appears all this method does is checking whether a block is a sign or not. 
Therefore you shouldn't need to iterate over all signs and stances in a switch but can use the `signs` tag which covers all `standing_signs` and `wall_signs` which again cover all stances of signs the platform provides. 
That is probably a more modern solution, assuming this is legacy code before block tags came in handy that much.

NotMyFault replied:
Amendment: 
This also relates to my other command with the BlockCategories.
Using them something like 
```java
private boolean isSign(BlockType blockType) {
        BlockType sign = BlockTypes.SIGN;
        BlockType wallSign = BlockTypes.WALL_SIGN;
        return blockType == sign || blockType == wallSign
            || BlockCategories.SIGNS.contains(blockType);
    }
```
would be possible.
Out of my mind I'm not quite sure but (FA)WE also contains a similar method that is public afaik. Could probably use that then.

Aurelien30000 replied:
Sound good, ty, I'll test it quickly.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/util/material/Materials.java:
```diff
@@ -8,7 +8,8 @@ private Materials() {
-	public static boolean isEmpty(Material material) {
-		return material == Material.AIR || material == Material.CAVE_AIR || material == Material.VOID_AIR;
+	public static boolean isLiquid(Material material) {
+		return material == Material.WATER || material == Material.LAVA;
```
Is this intended?

Aurelien30000 replied:
I remove ``#isEmpty`` method, useless, and added a util method ``#isLiquid``, already defined in several classes or piece of code.

NotMyFault reviewed (dismissed):
Did you take a look at MaterialSets yet? It appears to be referenced in methods quite often to check whether a targeted block matches a block range, but only uses legacy material directly referenced from the Material enum.
It would probably be better to use something like BlockCategories from WE or use tags the platform provides, instead of adding all new materials by hand.

Aurelien30000 committed:
Remove all unused imports

Aurelien30000 committed:
Replace ``#isSign`` method with a more reliable way

Aurelien30000 committed:
Use Worldedit alternatives when possible, Bukkit api is now only used for basic commands, to let users using a common format.

Aurelien30000 committed:
Remove Voxel undo and delegate work to FAWE.

Aurelien30000 committed:
Don't parse json text for signs, because it returns an adventure component as String.

Aurelien30000 committed:
Cleanup code:
- Add a few final fields
- Replace old code style prior Java 8
- Fix a minor performance issues

Aurelien30000 committed:
Remove useless allocation previously used for undo work.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SignOverwriteBrush.java:
```diff
@@ -335,6 +349,20 @@ private void resetStates() {
+		BlockType wallSign = BlockTypes.WALL_SIGN;
+		return blockType == sign || blockType == wallSign
+			|| BlockCategories.SIGNS.contains(blockType);
+	}
```
Could probably be replaced by `DeprecationUtil#isSign()`.

Aurelien30000 committed:
Replace ``isSign`` method by ``DeprecationUtil#isSign``.

NotMyFault reviewed (approved):
I'll approve this for now as it compiles, works and history isn't broken. Tho, this is minimal testing but so is my experience with (FA)VS.
I also requested a review from @SirYwell incase he wants to take a look over it, if not that's fine as well for such a big diff I guess.
